Import Ambr parser code

somtirtharoy opened the pull request.

somtirtharoy committed:
Initial commit for ambr parser

wip: adding in ambr parser for edd import

{wip} Fixed pre-commit errors

{wip} removed redundant code and added commants for better readability

somtirtharoy committed:
Add parser for parsing Ambr250 data for import into EDD

{wip} put ambr parser in a separate file, created new mixin for multi sheet ecel parsing, setup ambr parser testing code

{wip} added pandas in Pipfile and Pipfile.lock, put Ambrparser in its own file

Added mapper and codeto manipulate the ambr data using the GenericImportParser, wrote a simple test to check if parsed result is valid, addded Pipfile to root directory dso that tests can be run locally using IntelliJ, created a demo.py for testing

{wip} modified multisheetexcelparser to combine all imports as a single import call to the GenericImport parser parse_rows, modify the Measurement Mapper to have a map_data method that does alll the work, add drop nan values form the data

Made changes according to feedback from pull request

Modified test for ambr parser, removed large file from data folder

Added condition for renaming measurement types for fidelity with EDD

Add Layout for Ambr parser, modify handling of NaN values in the parser, make necesssary changes in testing the parser, rename sheet names in the Ambr test file

{wip} Add DefaultUnit class to make add mapping between measurement types and measurement units

somtirtharoy committed:
Add mapper for measurement types to units

- Add a class DefaultUnit that maps measurement tyeps to corresponding units
- Added admin class to configure the model from the admin panel
- optimized code to use the class instead of hardcoded map
- optmized code to create instance of Ambr data only once
- enable test to query database

somtirtharoy committed:
Database dump and update Pipfile.lock

Populate test databse with dump

- added dump from databse for the DefaultUnit table
- use the json file to load the data to populate the test database before running test for Ambr

Update Pipfile.lock

somtirtharoy committed:
Make changes according to feedback to pull request

- remove instance of df_parsed form code
- move DefaultUnit model from main module to edd.load module
- restructured code to move data manipultion code out of the parse function

somtirtharoy committed:
Configure test database using factories

- Add factories for measurement types and units
- Add test setup to create the default units in the test database
- make changes in parser to persist changes in the parsed result

somtirtharoy committed:
Add data configuration to map input measurement names to EDD names

- Add model to map any input measuremetn names to corresponding EDD names
- Add admin class to configure the model through the admin panel
- Add factory class to populate it in test databse before testing code
- Modify parser code to use the model rther than use hardcoded data
- Create migrations
- Delete defaultunit_data.json file from tests

Update Pipfile.lock

wmorrell commented on server/edd/load/migrations/0004_auto_20211005_0854.py:
```diff
@@ -0,0 +1,31 @@
+# Generated by Django 3.2.7 on 2021-10-05 15:54
```
Please update to include only one migration for the changes in the PR. The best way to do this is to:
1. run `python manage.py migrate load 0002_builtin_bootstrap` to roll back the existing migrations;
2. delete the migration files added here;
3. run `python manage.py makemigrations --name defaultunit load` to re-generate a single migration in the final model state;

somtirtharoy replied:
Done

wmorrell commented on server/edd/load/models.py:
```diff
@@ -152,3 +152,48 @@ class Meta:
+        verbose_name=_("Input Measurement Type"),
+    )
+
+    edd_type_name = VarCharField(
```
This should be a `ForeignKey` field to the actual type to use.

somtirtharoy replied:
this is a mapping between types so I want to keep this as a text field for the moment. I want to introduce this later with further improvements.

wmorrell replied:
Changing data models involves doing migrations, and excessive migration churn quickly becomes a problem in projects. Get the data model right up-front, because getting it wrong initially makes everything harder down the line. This needs to be a `ForeignKey` to the actual type. Don't just leave everything as strings.

wmorrell commented on server/edd/load/parsers/core.py:
```diff
@@ -641,6 +645,125 @@ def _raw_cell_value(self, cell):
+
+        # get the default units for measurements from the database
+        self.mes_unit_map = {}
+        du_obj = DefaultUnit.objects.all()
```
This is loading everything in this table into memory. This mapper should only load in the `DefaultUnit` instances that will apply to the current mapping operation, by doing a lookup based on having a matching `Protocol` (or the `DefaultUnit` specifying a blank/`None` `Protocol`), matching parser class (or the parser class specifying blank/`None`), and matching `MeasurementType`.

somtirtharoy replied:
Done

wmorrell commented on server/edd/load/parsers/core.py:
```diff
@@ -641,6 +645,125 @@ def _raw_cell_value(self, cell):
+        # get the mapping between input measurement type names and expected
+        # edd measurement type names from database
+        self.mtype_map = {}
+        mtype_obj = MeasurementNameTransform.objects.all()
```
Same as above, this is loading everything in the table into memory. Do a lookup with the specific names used, as they are used. This should probably also have a `Protocol` to constrain the results, as different reactor setups may export different names for the same things depending on how the operator(s) have set up the Ambr. Constraining by parser used should be included as well.

somtirtharoy replied:
Done

wmorrell commented on server/edd/load/tests/factory.py:
```diff
@@ -52,3 +54,35 @@ class Meta:
     mime_type = factory.Faker("mime_type")
+
+
+class MeasurementTypeFactory(factory.django.DjangoModelFactory):
```
There's already a `MeasurementTypeFactory` in `main.tests.factory`, should use that one instead, or document a justification for having a separate one here.

somtirtharoy replied:
Done

wmorrell commented on server/edd/load/tests/factory.py:
```diff
@@ -52,3 +54,35 @@ class Meta:
+    type_name = ""
+
+
+class MeasurementUnitFactory(factory.django.DjangoModelFactory):
```
Same here, there's already a `UnitFactory` in `main.tests.factory`.

somtirtharoy replied:
Done

wmorrell commented on server/edd/load/tests/factory.py:
```diff
@@ -52,3 +54,35 @@ class Meta:
+    class Meta:
+        model = models.DefaultUnit
+
+    measurement_type = factory.SubFactory(LayoutFactory)
```
The subfactory here should be `MeasurementTypeFactory`, not `LayoutFactory`.

somtirtharoy replied:
Done

wmorrell commented on server/edd/load/tests/factory.py:
```diff
@@ -52,3 +54,35 @@ class Meta:
+        model = models.DefaultUnit
+
+    measurement_type = factory.SubFactory(LayoutFactory)
+    unit = factory.SubFactory(LayoutFactory)
```
The subfactory here should be `UnitFactory`, not `LayoutFactory`.

somtirtharoy replied:
Done

wmorrell commented on server/main/migrations/0002_defaultunit.py:
```diff
@@ -0,0 +1,55 @@
+# Generated by Django 2.2.23 on 2021-09-22 14:41
```
This migration file should not exist; the models are in `edd.load` app now, not in `main` app.

somtirtharoy replied:
Done

wmorrell commented on server/main/migrations/0003_delete_defaultunit.py:
```diff
@@ -0,0 +1,14 @@
+# Generated by Django 3.2.7 on 2021-10-05 15:46
```
This migration file should not exist, either. Steps are similar to fixing the migrations for `edd.load` app:
1. run `python manage.py migrate main 0001_edd_2_7` to roll back;
2. delete these files and update commits to remove them from the pull request;

somtirtharoy replied:
Done

somtirtharoy commented:
I am on it. Will get this to you tonight! Thanks so much!

On Thu, Oct 7, 2021 at 2:13 PM wmorrell ***@***.***> wrote:

> ***@***.**** requested changes on this pull request.
> ------------------------------
>
> In server/edd/load/migrations/0004_auto_20211005_0854.py
> <https://github.com/JBEI/edd/pull/27#discussion_r724493084>:
>
> > @@ -0,0 +1,31 @@
> +# Generated by Django 3.2.7 on 2021-10-05 15:54
>
> Please update to include only one migration for the changes in the PR. The
> best way to do this is to:
>
>    1. run python manage.py migrate load 0002_builtin_bootstrap to roll
>    back the existing migrations;
>    2. delete the migration files added here;
>    3. run python manage.py makemigrations --name defaultunit load to
>    re-generate a single migration in the final model state;
>
> ------------------------------
>
> In server/edd/load/models.py
> <https://github.com/JBEI/edd/pull/27#discussion_r724494363>:
>
> > +            "id": self.pk,
> +            "type_name": self.measurement_type.type_name,
> +            "unit_name": self.unit.unit_name,
> +        }
> +
> +
> +class MeasurementNameTransform(models.Model):
> +    class Meta:
> +        db_table = "measurement_name_transform"
> +
> +    input_type_name = VarCharField(
> +        help_text=_("Name of this Measurement Type in input."),
> +        verbose_name=_("Input Measurement Type"),
> +    )
> +
> +    edd_type_name = VarCharField(
>
> This should be a ForeignKey field to the actual type to use.
> ------------------------------
>
> In server/edd/load/parsers/core.py
> <https://github.com/JBEI/edd/pull/27#discussion_r724519152>:
>
> > +
> +class MeasurementMapper:
> +
> +    loa_name: str
> +    df: pd.DataFrame
> +    units: Dict
> +    mes_unit_map: Dict
> +    mtype_map = Dict
> +
> +    def __init__(self, name=None, df=None):
> +        self.loa_name = name
> +        self.df = df
> +
> +        # get the default units for measurements from the database
> +        self.mes_unit_map = {}
> +        du_obj = DefaultUnit.objects.all()
>
> This is loading everything in this table into memory. This mapper should
> only load in the DefaultUnit instances that will apply to the current
> mapping operation, by doing a lookup based on having a matching Protocol
> (or the DefaultUnit specifying a blank/None Protocol), matching parser
> class (or the parser class specifying blank/None), and matching
> MeasurementType.
> ------------------------------
>
> In server/edd/load/parsers/core.py
> <https://github.com/JBEI/edd/pull/27#discussion_r724525528>:
>
> > +    mtype_map = Dict
> +
> +    def __init__(self, name=None, df=None):
> +        self.loa_name = name
> +        self.df = df
> +
> +        # get the default units for measurements from the database
> +        self.mes_unit_map = {}
> +        du_obj = DefaultUnit.objects.all()
> +        for obj in du_obj:
> +            self.mes_unit_map[obj.measurement_type.type_name] = obj.unit.unit_name
> +
> +        # get the mapping between input measurement type names and expected
> +        # edd measurement type names from database
> +        self.mtype_map = {}
> +        mtype_obj = MeasurementNameTransform.objects.all()
>
> Same as above, this is loading everything in the table into memory. Do a
> lookup with the specific names used, as they are used. This should probably
> also have a Protocol to constrain the results, as different reactor
> setups may export different names for the same things depending on how the
> operator(s) have set up the Ambr. Constraining by parser used should be
> included as well.
> ------------------------------
>
> In server/edd/load/tests/factory.py
> <https://github.com/JBEI/edd/pull/27#discussion_r724527326>:
>
> > @@ -52,3 +54,35 @@ class Meta:
>
>      layout = factory.SubFactory(LayoutFactory)
>      mime_type = factory.Faker("mime_type")
> +
> +
> +class MeasurementTypeFactory(factory.django.DjangoModelFactory):
>
> There's already a MeasurementTypeFactory in main.tests.factory, should
> use that one instead, or document a justification for having a separate one
> here.
> ------------------------------
>
> In server/edd/load/tests/factory.py
> <https://github.com/JBEI/edd/pull/27#discussion_r724527830>:
>
> > @@ -52,3 +54,35 @@ class Meta:
>
>      layout = factory.SubFactory(LayoutFactory)
>      mime_type = factory.Faker("mime_type")
> +
> +
> +class MeasurementTypeFactory(factory.django.DjangoModelFactory):
> +    class Meta:
> +        model = main_models.MeasurementType
> +        django_get_or_create = ("type_name",)
> +
> +    type_name = ""
> +
> +
> +class MeasurementUnitFactory(factory.django.DjangoModelFactory):
>
> Same here, there's already a UnitFactory in main.tests.factory.
> ------------------------------
>
> In server/edd/load/tests/factory.py
> <https://github.com/JBEI/edd/pull/27#discussion_r724528491>:
>
> > +    type_name = ""
> +
> +
> +class MeasurementUnitFactory(factory.django.DjangoModelFactory):
> +    class Meta:
> +        model = main_models.MeasurementUnit
> +        django_get_or_create = ("unit_name",)
> +
> +    unit_name = ""
> +
> +
> +class DefaultUnitFactory(factory.django.DjangoModelFactory):
> +    class Meta:
> +        model = models.DefaultUnit
> +
> +    measurement_type = factory.SubFactory(LayoutFactory)
>
> The subfactory here should be MeasurementTypeFactory, not LayoutFactory.
> ------------------------------
>
> In server/edd/load/tests/factory.py
> <https://github.com/JBEI/edd/pull/27#discussion_r724529564>:
>
> > +
> +
> +class MeasurementUnitFactory(factory.django.DjangoModelFactory):
> +    class Meta:
> +        model = main_models.MeasurementUnit
> +        django_get_or_create = ("unit_name",)
> +
> +    unit_name = ""
> +
> +
> +class DefaultUnitFactory(factory.django.DjangoModelFactory):
> +    class Meta:
> +        model = models.DefaultUnit
> +
> +    measurement_type = factory.SubFactory(LayoutFactory)
> +    unit = factory.SubFactory(LayoutFactory)
>
> The subfactory here should be UnitFactory, not LayoutFactory.
> ------------------------------
>
> In server/main/migrations/0002_defaultunit.py
> <https://github.com/JBEI/edd/pull/27#discussion_r724530385>:
>
> > @@ -0,0 +1,55 @@
> +# Generated by Django 2.2.23 on 2021-09-22 14:41
>
> This migration file should not exist; the models are in edd.load app now,
> not in main app.
> ------------------------------
>
> In server/main/migrations/0003_delete_defaultunit.py
> <https://github.com/JBEI/edd/pull/27#discussion_r724530733>:
>
> > @@ -0,0 +1,14 @@
> +# Generated by Django 3.2.7 on 2021-10-05 15:46
>
> This migration file should not exist, either. Steps are similar to fixing
> the migrations for edd.load app:
>
>    1. run python manage.py migrate main 0001_edd_2_7 to roll back;
>    2. delete these files and update commits to remove them from the pull
>    request;
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/JBEI/edd/pull/27#pullrequestreview-774338440>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKDGPCAB7AOICFQOVYTZU7LUFYEPNANCNFSM5FQNBYZQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

somtirtharoy committed:
Make changes according to feedback from pull request

- deleted migrations to consolidate all migrations under one migration
- Removed redundant factories that already exist in main fro measurement type and unit
- changed code to load only record for defaultuit and measurement transform fileterd by measurement type, name, parser and protocol
- changed the class ofr the subfactory in default factory to measuremetn type and unit

somtirtharoy commented:
I pushed the changes, please check.

On Thu, Oct 7, 2021 at 4:15 PM Somtirtha Roy ***@***.***> wrote:

> I am on it. Will get this to you tonight! Thanks so much!
>
> On Thu, Oct 7, 2021 at 2:13 PM wmorrell ***@***.***> wrote:
>
>> ***@***.**** requested changes on this pull request.
>> ------------------------------
>>
>> In server/edd/load/migrations/0004_auto_20211005_0854.py
>> <https://github.com/JBEI/edd/pull/27#discussion_r724493084>:
>>
>> > @@ -0,0 +1,31 @@
>> +# Generated by Django 3.2.7 on 2021-10-05 15:54
>>
>> Please update to include only one migration for the changes in the PR.
>> The best way to do this is to:
>>
>>    1. run python manage.py migrate load 0002_builtin_bootstrap to roll
>>    back the existing migrations;
>>    2. delete the migration files added here;
>>    3. run python manage.py makemigrations --name defaultunit load to
>>    re-generate a single migration in the final model state;
>>
>> ------------------------------
>>
>> In server/edd/load/models.py
>> <https://github.com/JBEI/edd/pull/27#discussion_r724494363>:
>>
>> > +            "id": self.pk,
>> +            "type_name": self.measurement_type.type_name,
>> +            "unit_name": self.unit.unit_name,
>> +        }
>> +
>> +
>> +class MeasurementNameTransform(models.Model):
>> +    class Meta:
>> +        db_table = "measurement_name_transform"
>> +
>> +    input_type_name = VarCharField(
>> +        help_text=_("Name of this Measurement Type in input."),
>> +        verbose_name=_("Input Measurement Type"),
>> +    )
>> +
>> +    edd_type_name = VarCharField(
>>
>> This should be a ForeignKey field to the actual type to use.
>> ------------------------------
>>
>> In server/edd/load/parsers/core.py
>> <https://github.com/JBEI/edd/pull/27#discussion_r724519152>:
>>
>> > +
>> +class MeasurementMapper:
>> +
>> +    loa_name: str
>> +    df: pd.DataFrame
>> +    units: Dict
>> +    mes_unit_map: Dict
>> +    mtype_map = Dict
>> +
>> +    def __init__(self, name=None, df=None):
>> +        self.loa_name = name
>> +        self.df = df
>> +
>> +        # get the default units for measurements from the database
>> +        self.mes_unit_map = {}
>> +        du_obj = DefaultUnit.objects.all()
>>
>> This is loading everything in this table into memory. This mapper should
>> only load in the DefaultUnit instances that will apply to the current
>> mapping operation, by doing a lookup based on having a matching Protocol
>> (or the DefaultUnit specifying a blank/None Protocol), matching parser
>> class (or the parser class specifying blank/None), and matching
>> MeasurementType.
>> ------------------------------
>>
>> In server/edd/load/parsers/core.py
>> <https://github.com/JBEI/edd/pull/27#discussion_r724525528>:
>>
>> > +    mtype_map = Dict
>> +
>> +    def __init__(self, name=None, df=None):
>> +        self.loa_name = name
>> +        self.df = df
>> +
>> +        # get the default units for measurements from the database
>> +        self.mes_unit_map = {}
>> +        du_obj = DefaultUnit.objects.all()
>> +        for obj in du_obj:
>> +            self.mes_unit_map[obj.measurement_type.type_name] = obj.unit.unit_name
>> +
>> +        # get the mapping between input measurement type names and expected
>> +        # edd measurement type names from database
>> +        self.mtype_map = {}
>> +        mtype_obj = MeasurementNameTransform.objects.all()
>>
>> Same as above, this is loading everything in the table into memory. Do a
>> lookup with the specific names used, as they are used. This should probably
>> also have a Protocol to constrain the results, as different reactor
>> setups may export different names for the same things depending on how the
>> operator(s) have set up the Ambr. Constraining by parser used should be
>> included as well.
>> ------------------------------
>>
>> In server/edd/load/tests/factory.py
>> <https://github.com/JBEI/edd/pull/27#discussion_r724527326>:
>>
>> > @@ -52,3 +54,35 @@ class Meta:
>>
>>      layout = factory.SubFactory(LayoutFactory)
>>      mime_type = factory.Faker("mime_type")
>> +
>> +
>> +class MeasurementTypeFactory(factory.django.DjangoModelFactory):
>>
>> There's already a MeasurementTypeFactory in main.tests.factory, should
>> use that one instead, or document a justification for having a separate one
>> here.
>> ------------------------------
>>
>> In server/edd/load/tests/factory.py
>> <https://github.com/JBEI/edd/pull/27#discussion_r724527830>:
>>
>> > @@ -52,3 +54,35 @@ class Meta:
>>
>>      layout = factory.SubFactory(LayoutFactory)
>>      mime_type = factory.Faker("mime_type")
>> +
>> +
>> +class MeasurementTypeFactory(factory.django.DjangoModelFactory):
>> +    class Meta:
>> +        model = main_models.MeasurementType
>> +        django_get_or_create = ("type_name",)
>> +
>> +    type_name = ""
>> +
>> +
>> +class MeasurementUnitFactory(factory.django.DjangoModelFactory):
>>
>> Same here, there's already a UnitFactory in main.tests.factory.
>> ------------------------------
>>
>> In server/edd/load/tests/factory.py
>> <https://github.com/JBEI/edd/pull/27#discussion_r724528491>:
>>
>> > +    type_name = ""
>> +
>> +
>> +class MeasurementUnitFactory(factory.django.DjangoModelFactory):
>> +    class Meta:
>> +        model = main_models.MeasurementUnit
>> +        django_get_or_create = ("unit_name",)
>> +
>> +    unit_name = ""
>> +
>> +
>> +class DefaultUnitFactory(factory.django.DjangoModelFactory):
>> +    class Meta:
>> +        model = models.DefaultUnit
>> +
>> +    measurement_type = factory.SubFactory(LayoutFactory)
>>
>> The subfactory here should be MeasurementTypeFactory, not LayoutFactory.
>> ------------------------------
>>
>> In server/edd/load/tests/factory.py
>> <https://github.com/JBEI/edd/pull/27#discussion_r724529564>:
>>
>> > +
>> +
>> +class MeasurementUnitFactory(factory.django.DjangoModelFactory):
>> +    class Meta:
>> +        model = main_models.MeasurementUnit
>> +        django_get_or_create = ("unit_name",)
>> +
>> +    unit_name = ""
>> +
>> +
>> +class DefaultUnitFactory(factory.django.DjangoModelFactory):
>> +    class Meta:
>> +        model = models.DefaultUnit
>> +
>> +    measurement_type = factory.SubFactory(LayoutFactory)
>> +    unit = factory.SubFactory(LayoutFactory)
>>
>> The subfactory here should be UnitFactory, not LayoutFactory.
>> ------------------------------
>>
>> In server/main/migrations/0002_defaultunit.py
>> <https://github.com/JBEI/edd/pull/27#discussion_r724530385>:
>>
>> > @@ -0,0 +1,55 @@
>> +# Generated by Django 2.2.23 on 2021-09-22 14:41
>>
>> This migration file should not exist; the models are in edd.load app
>> now, not in main app.
>> ------------------------------
>>
>> In server/main/migrations/0003_delete_defaultunit.py
>> <https://github.com/JBEI/edd/pull/27#discussion_r724530733>:
>>
>> > @@ -0,0 +1,14 @@
>> +# Generated by Django 3.2.7 on 2021-10-05 15:46
>>
>> This migration file should not exist, either. Steps are similar to fixing
>> the migrations for edd.load app:
>>
>>    1. run python manage.py migrate main 0001_edd_2_7 to roll back;
>>    2. delete these files and update commits to remove them from the pull
>>    request;
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/JBEI/edd/pull/27#pullrequestreview-774338440>, or
>> unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AKDGPCAB7AOICFQOVYTZU7LUFYEPNANCNFSM5FQNBYZQ>
>> .
>> Triage notifications on the go with GitHub Mobile for iOS
>> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
>> or Android
>> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>>
>>
>

wmorrell commented on docker/edd/core/Pipfile.lock:
```diff
@@ -780,6 +781,42 @@
             ],
             "version": "==1.0.2"
         },
+        "numpy": {
```
The lock file here is re-installing `numpy` over top of the version that's already installed in the base `scikit-learn` image. I believe this is because of adding `pandas`. You should not need to include `pandas` at all and change the installed packages relative to what's already in trunk.

wmorrell commented on server/edd/load/migrations/0002_builtin_bootstrap.py:
```diff
@@ -6,6 +6,7 @@ def bootstrap(apps, schema_editor):
     Layout = apps.get_model("load", "Layout")
     LAYOUT_GENERIC = Layout.objects.create(name="Generic", description="")
     LAYOUT_SKYLINE = Layout.objects.create(name="Skyline", description="")
+    LAYOUT_AMBR = Layout.objects.create(name="Ambr", description="")
```
Since you're adding models and generating a new migration anyway, the bootstrap for the new parser should go in with the new migration `0003_defaultunit` instead of editing this one and manually inserting into existing deployments.

wmorrell commented on server/edd/load/parsers/core.py:
```diff
@@ -641,6 +646,122 @@ def _raw_cell_value(self, cell):
         return val
 
 
+class MultiSheetExcelParserMixin:
```
This class is still doing too much. The functionality of the `MeasurementMapper`, selecting pairs of columns, etc. needs to be in `AmbrExcelParser`. A parser that's dealing with multiple Excel sheets should *only* be dealing with looping over all sheets, and calling a function to process the rows for each one. Thus, the `parse(file)` function should look something like this:

```python
def parse(self, file):
    wb = load_workbook(file, read_only=True, data_only=True)
    for ws in wb.worksheets:
        self._parse_sheet_rows(ws)
    return self._parse_result
```

It's then the responsibility for any classes using `MultiSheetExcelParserMixin` (including `AmbrExcelParser`) to define a `_parse_sheet_rows(self, ws)` method that will process the sheet, and a `_parse_result` property that takes all the collected info from all the sheets processed, and returns a `ParseResult` instance.

wmorrell commented on server/edd/load/parsers/generic.py:
```diff
@@ -95,3 +95,5 @@ class GenericExcelParser(ExcelParserMixin, GenericImportParser):
 
 class GenericCsvParser(CsvParserMixin, GenericImportParser):
     pass
+
```
Make sure you have `pre-commit` set up to run. These empty lines should have been deleted by the repo's hooks.

wmorrell commented on server/edd/load/tests/test_ambr_parser.py:
```diff
@@ -0,0 +1,172 @@
+from .factory import DefaultUnitFactory, MeasurementNameTransformFactory
+
+
+@pytest.mark.django_db
```
The `pytest.mark.django_db` annotation is for marking tests. It has no function for items not called by `pytest`. See: https://pytest-django.readthedocs.io/en/latest/helpers.html#pytest.mark.django_db

somtirtharoy committed:
Fix test-resolver tests failing after adding ambr parser code

- removed the get_or_create attribute in the Meta class for MeasuremetnTypeFactory

somtirtharoy committed:
Restructure ambr parser code

- Restructure code to put column manipulation code in AmbrExcelParser class and make parse function in the MultiSheetExcelParserMixin class handle only readin the excel sheet and seding back the result

somtirtharoy committed:
Modify models and consolidate migrations into one migration

- Changed field in MeasurementNameTransform from varchar to foreighnkey type and in factories
- consolidated all migrations, added bootstrap data for new models
- moved the layout and parser from 0002_builtin_bootstrap to 0003_defaultunit
- modified test to use bootstrap data instead of factories
- made modifications in map_data code

somtirtharoy committed:
Replaced pandas with openpyxl in the ambr parser

- replaced pandas with openpyxl
- modifoed test file, rewrote tests
- added code to decimate only for large data
- rewrote code to drop NaN values
